pkg/server/plugin/upstreamauthority/vault/vault_client.go

```diff
@@ -112,47 +121,55 @@ func NewClientConfig(cp *ClientParams, logger hclog.Logger) (*ClientConfig, erro
 }
 
 // NewAuthenticatedClient returns a new authenticated vault client with given authentication method
-func (c *ClientConfig) NewAuthenticatedClient(method AuthMethod) (*Client, error) {
+func (c *ClientConfig) NewAuthenticatedClient(method AuthMethod) (*Client, bool, error) {
 	config := vapi.DefaultConfig()
 	config.Address = c.clientParams.VaultAddr
 	if c.clientParams.MaxRetries != nil {
 		config.MaxRetries = *c.clientParams.MaxRetries
 	}
 
 	if err := c.configureTLS(config); err != nil {
-		return nil, err
+		return nil, false, err
 	}
 	vc, err := vapi.NewClient(config)
 	if err != nil {
-		return nil, err
+		return nil, false, err
 	}
 
 	client := &Client{
 		vaultClient:  vc,
 		clientParams: c.clientParams,
 	}
 
+	var ts TokenStatus
 	switch method {
 	case TOKEN:
+		sec, err := client.LookupSelf(c.clientParams.Token)
+		if err != nil {
+			return nil, false, err
+		}
+		if sec == nil {
+			return nil, false, errors.New("lookup self response is nil")
+		}
+		ts, err = handleRenewToken(vc, sec, c.Logger)
+		if err != nil {
+			return nil, false, err
+		}
```

Comment:
The same thing is repeated in all the cases. Maybe this can be done at the end of the function before returning?

Reply:
Thank you! Fix it to be DRY.